File: NearestNeighborTSP.java
import java.io.*;
import java.util.*;

public class NearestNeighborTSP {
    private static final double INF = Double.MAX_VALUE;

    // CSV file reader. Returns a map containing the distance matrix and city names
    public static Map<String, Object> readCSV(String filePath) throws IOException {
        List<String> cityNames = new ArrayList<>(); // List to store city names
        Map<String, Integer> cityIndexMap = new HashMap<>(); // Map to store city name to index mapping
        String line; // Variable to hold each line read from the file

        // Try-with-resources to ensure the BufferedReader is closed after use
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            line = br.readLine(); // Read the header line
            // Check if the header line is present and correctly formatted
            if (line == null || !line.trim().equalsIgnoreCase("City1,City2,Distance(km)")) {
                throw new IOException("CSV file must have header: City1,City2,Distance(km)");
            }

            // Read each line of the CSV file
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(","); // Split line into components
                // Validate the row format
                if (parts.length != 3) {
                    throw new IOException("Invalid row format: " + line);
                }

                String city1 = parts[0].trim(); // First city
                String city2 = parts[1].trim(); // Second city
                double distance = Double.parseDouble(parts[2].trim()); // Distance between the cities

                // Add the first city to the city list and map if it doesn't exist
                if (!cityIndexMap.containsKey(city1)) {
                    cityIndexMap.put(city1, cityNames.size()); // Map city to its index
                    cityNames.add(city1); // Add city to list
                }
                // Add the second city to the city list and map if it doesn't exist
                if (!cityIndexMap.containsKey(city2)) {
                    cityIndexMap.put(city2, cityNames.size()); // Map city to its index
                    cityNames.add(city2); // Add city to list
                }
            }
        }

        int n = cityNames.size(); // Number of unique cities
        double[][] distanceMatrix = new double[n][n]; // Initialize the distance matrix
        for (int i = 0; i < n; i++) {
            Arrays.fill(distanceMatrix[i], INF); // Fill with INF to represent no direct connection
            distanceMatrix[i][i] = 0; // Distance to self is always 0
        }

        // Re-read the file to populate the distance matrix
        try (BufferedReader br = new BufferedReader(new FileReader(filePath))) {
            br.readLine(); // Skip the header line
            while ((line = br.readLine()) != null) {
                String[] parts = line.split(","); // Split line into components
                String city1 = parts[0].trim(); // First city
                String city2 = parts[1].trim(); // Second city
                double distance = Double.parseDouble(parts[2].trim()); // Distance between the cities

                int index1 = cityIndexMap.get(city1); // Retrieve index of the first city
                int index2 = cityIndexMap.get(city2); // Retrieve index of the second city

                // Populate the symmetric distance matrix with the distance
                distanceMatrix[index1][index2] = distance;
                distanceMatrix[index2][index1] = distance;
            }
        }

        // Prepare the result map to return
        Map<String, Object> result = new HashMap<>();
        result.put("cityNames", cityNames); // Add city names to the result
        result.put("distanceMatrix", distanceMatrix); // Add distance matrix to the result

        return result; // Return the result map
    }

    // Calculates the total distance of a given tour
    public static double calculateTourDistance(double[][] distanceMatrix, List<Integer> tour) {
        // Initialize total distance to 0
        double totalDistance = 0;
        
        // Loop through the tour list to calculate the distance between consecutive cities
        for (int i = 0; i < tour.size() - 1; i++) {
            // Add the distance from the current city to the next city in the tour
            totalDistance += distanceMatrix[tour.get(i)][tour.get(i + 1)];
        }
        
        // Add the distance from the last city back to the starting city to complete the tour
        totalDistance += distanceMatrix[tour.get(tour.size() - 1)][tour.get(0)];
        
        // Round the total distance to two decimal places and return it
        return roundToTwoDecimals(totalDistance);
    }

    // Rounds double to 2 decimal places
    public static double roundToTwoDecimals(double value) {
        return Math.round(value * 100.0) / 100.0;
    }

    // Nearest Neighbor Algorithm
    public static List<Integer> nearestNeighbor(double[][] distanceMatrix, int startCity) {
        int n = distanceMatrix.length;
        List<Integer> tour = new ArrayList<>();
        boolean[] visited = new boolean[n];
        int currentCity = startCity;

        visited[currentCity] = true;
        tour.add(currentCity);

        for (int i = 1; i < n; i++) {
            double nearestDistance = INF;
            int nearestCity = -1;

            // Find the nearest unvisited city
            for (int j = 0; j < n; j++) {
                if (!visited[j] && distanceMatrix[currentCity][j] < nearestDistance) {
                    nearestDistance = distanceMatrix[currentCity][j];
                    nearestCity = j;
                }
            }

            visited[nearestCity] = true;
            tour.add(nearestCity);
            currentCity = nearestCity;
        }

        return tour;
    }

    // Prints tour with city names
    public static void printTourWithCityNames(List<Integer> tour, List<String> cityNames) {
        for (int i : tour) {
            System.out.print(cityNames.get(i) + " -> ");
        }
        System.out.println(cityNames.get(tour.get(0))); // Returns to start
    }

    public static void main(String[] args) {
        // File path to CSV file
        String filePath = "DistanceBetweenEuropeanCities.csv";

        try {
            // Read data from CSV file
            Map<String, Object> data = readCSV(filePath);
            List<String> cityNames = (List<String>) data.get("cityNames");
            double[][] distanceMatrix = (double[][]) data.get("distanceMatrix");

            int n = distanceMatrix.length;

            double bestDistance = INF;
            List<Integer> bestTour = null;
            int bestStartCity = -1;

            // Computation time begins
            long startTime = System.nanoTime();

            // Tour optimization from each city
            for (int startCity = 0; startCity < n; startCity++) {
                List<Integer> tour = nearestNeighbor(distanceMatrix, startCity);
                double tourDistance = calculateTourDistance(distanceMatrix, tour);

                System.out.println("\nTour Starting from " + cityNames.get(startCity) + ":");
                printTourWithCityNames(tour, cityNames);
                System.out.println("Distance: " + String.format("%.2f", tourDistance));

                // Update the best tour if this one is better
                if (tourDistance < bestDistance) {
                    bestDistance = tourDistance;
                    bestTour = tour;
                    bestStartCity = startCity;
                }
            }

            // Computation time ends
            long endTime = System.nanoTime();

            // Elapsed time in milliseconds
            double elapsedTime = (endTime - startTime) / 1_000_000.0;

            // Best overall tour
            System.out.println("\nBest Tour Starting from " + cityNames.get(bestStartCity) + ":");
            printTourWithCityNames(bestTour, cityNames);
            System.out.println("Best Distance: " + String.format("%.2f", bestDistance));

            // Time taken to calculate
            System.out.printf("\nComputation Time: %.2f milliseconds\n", elapsedTime);

            // Hardware + Software declaration
            System.out.println("\nOperating System: " + System.getProperty("os.name"));
            System.out.println("OS Version: " + System.getProperty("os.version"));
            System.out.println("OS Architecture: " + System.getProperty("os.arch"));
            System.out.println("User Name: " + System.getProperty("user.name"));
            System.out.println("Java Version: " + System.getProperty("java.version"));
            System.out.println("Java Vendor: " + System.getProperty("java.vendor"));
        } catch (IOException e) {
            System.err.println("Error reading the CSV file: " + e.getMessage());
        }
    }
}

File: NearestNeighborTSP.py
import csv
import random
import math
import time
import platform

INF = float('inf')

def read_csv(file_path):
    """
    Reads the CSV file and constructs a distance matrix and city names list.

    The CSV file is expected to have the following format:
    City1,City2,Distance(km)
    CityName1,CityName2,DistanceInKilometers
    CityName3,CityName4,DistanceInKilometers
    ...

    The function returns a tuple containing the city names list and the distance matrix.
    The city names list is a list of strings, where each string is the name of a city.
    The distance matrix is a 2D list of floats, where distance_matrix[i][j] is the distance in kilometers between city i and city j.
    If there is no direct connection between city i and city j, then distance_matrix[i][j] is infinity.
    """
    city_names = []
    city_index_map = {}
    edges = []

    with open(file_path, 'r') as csvfile:
        reader = csv.reader(csvfile)
        header = next(reader)
        if header != ["City1", "City2", "Distance(km)"]:
            raise ValueError("CSV file must have header: City1,City2,Distance(km)")

        for row in reader:
            # For each row in the CSV file, extract the city names and the distance between them
            city1, city2, distance = row[0].strip(), row[1].strip(), float(row[2].strip())

            # If city1 is not already in the city_index_map, add it and its index to the map
            if city1 not in city_index_map:
                city_index_map[city1] = len(city_names)
                city_names.append(city1)

            # If city2 is not already in the city_index_map, add it and its index to the map
            if city2 not in city_index_map:
                city_index_map[city2] = len(city_names)
                city_names.append(city2)

            # Add the edge between the two cities to the edges list
            edges.append((city1, city2, distance))

    # Create a distance matrix with the same number of rows and columns as the number of cities
    n = len(city_names)
    distance_matrix = [[INF] * n for _ in range(n)]

    # For each city, set the distance to itself to 0
    for i in range(n):
        distance_matrix[i][i] = 0

    # For each edge, set the distance between the two cities to the given distance
    for city1, city2, distance in edges:
        index1 = city_index_map[city1]
        index2 = city_index_map[city2]
        distance_matrix[index1][index2] = distance
        distance_matrix[index2][index1] = distance

    return city_names, distance_matrix

def calculate_tour_distance(distance_matrix, tour):
    """
    Calculates the total distance of a given tour.
    """
    total_distance = 0
    for i in range(len(tour) - 1):
        total_distance += distance_matrix[tour[i]][tour[i + 1]]
    total_distance += distance_matrix[tour[-1]][tour[0]]  # Return to start
    return round(total_distance, 2)

def nearest_neighbor(distance_matrix, start_city):
    """
    Implements the Nearest Neighbor algorithm.
    """
    n = len(distance_matrix)
    tour = [start_city]
    visited = [False] * n
    visited[start_city] = True

    current_city = start_city
    for _ in range(n - 1):
        nearest_distance = INF
        nearest_city = -1

        for next_city in range(n):
            if not visited[next_city] and distance_matrix[current_city][next_city] < nearest_distance:
                nearest_distance = distance_matrix[current_city][next_city]
                nearest_city = next_city

        visited[nearest_city] = True
        tour.append(nearest_city)
        current_city = nearest_city

    return tour

def print_tour_with_city_names(tour, city_names):
    """
    Prints the tour with city names.
    """
    tour_names = [city_names[city] for city in tour]
    print(" -> ".join(tour_names) + f" -> {city_names[tour[0]]}")

def main():
    file_path = "DistanceBetweenEuropeanCities.csv"

    try:
        city_names, full_distance_matrix = read_csv(file_path)
        total_cities = len(city_names)

        # Ask the user for the number of cities
        while True:
            try:
                num_cities = int(input(f"Enter the number of cities to use (2-{min(50, total_cities)}): "))
                if 2 <= num_cities <= min(50, total_cities):
                    break
                else:
                    print(f"Please enter a valid number between 2 and {min(50, total_cities)}.")
            except ValueError:
                print("Invalid input. Please enter an integer.")

        # Select a subset of cities
        selected_indices = sorted(random.sample(range(total_cities), num_cities))
        selected_city_names = [city_names[i] for i in selected_indices]

        # Create a reduced distance matrix
        distance_matrix = [[full_distance_matrix[i][j] for j in selected_indices] for i in selected_indices]

        # Run the Nearest Neighbor Algorithm
        n = len(distance_matrix)
        best_distance = INF
        best_tour = None
        best_start_city = -1

        start_time = time.time()

        for start_city in range(n):
            tour = nearest_neighbor(distance_matrix, start_city)
            tour_distance = calculate_tour_distance(distance_matrix, tour)

            print(f"\nTour Starting from {selected_city_names[start_city]}:")
            print_tour_with_city_names(tour, selected_city_names)
            print(f"Distance: {tour_distance:.2f} km")

            if tour_distance < best_distance:
                best_distance = tour_distance
                best_tour = tour
                best_start_city = start_city

        end_time = time.time()
        elapsed_time = (end_time - start_time) * 1000

        print(f"\nBest Tour Starting from {selected_city_names[best_start_city]}:")
        print_tour_with_city_names(best_tour, selected_city_names)
        print(f"Best Distance: {best_distance:.2f} km")

        print(f"\nComputation Time: {elapsed_time:.2f} milliseconds")

        print("\nOperating System:", platform.system())
        print("OS Version:", platform.version())
        print("OS Architecture:", platform.architecture()[0])
        print("Python Version:", platform.python_version())

    except Exception as e:
        print("Error:", str(e))

if __name__ == "__main__":
    main()

